nighthawk/test/sequencer_test.cc

```diff
@@ -0,0 +1,108 @@
+#include <chrono>
+
+#include "gtest/gtest.h"
+
+#include "test/test_common/simulated_time_system.h"
+
+#include "common/api/api_impl.h"
+#include "common/common/thread_impl.h"
+#include "common/event/dispatcher_impl.h"
+#include "common/stats/isolated_store_impl.h"
+
+#include "nighthawk/common/exception.h"
+
+#include "nighthawk/source/common/rate_limiter_impl.h"
+#include "nighthawk/source/common/sequencer_impl.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+
+class SequencerTest : public testing::Test {
+public:
+  SequencerTest()
+      : api_(1000ms /*flush interval*/, thread_factory_, store_, time_system_),
+        dispatcher_(api_.allocateDispatcher()), callback_test_count_(0), frequency_(10_Hz),
+        interval_(std::chrono::duration_cast<std::chrono::milliseconds>(frequency_.interval())),
+        times_(5), rate_limiter_(time_system_, frequency_),
+        sequencer_target_(std::bind(&SequencerTest::callback_test, this, std::placeholders::_1)),
+        clock_updates_(0) {}
+
+  void updateClock() {
+    time_system_.setMonotonicTime((clock_updates_ * interval_) + interval_);
+    clock_updates_++;
+  }
+
+  bool callback_test(std::function<void()> f) {
+    callback_test_count_++;
+    f();
+    return true;
+  }
+  bool timeout_test(std::function<void()> /* f */) {
+    callback_test_count_++;
+    // We don't call f(); which will cause the sequencer to think
+    // there's outstanding work.
+    return true;
+  }
+
+  Envoy::Thread::ThreadFactoryImplPosix thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  Envoy::Event::SimulatedTimeSystem time_system_;
+  Envoy::Api::Impl api_;
+  Envoy::Event::DispatcherPtr dispatcher_;
+  int callback_test_count_;
+  const Frequency frequency_;
+  const std::chrono::milliseconds interval_;
+  const uint64_t times_;
+  LinearRateLimiter rate_limiter_;
```

Comment:
Should we have a `MockRateLimiter`? Would mocking `dispacher_` provide any value?

Reply:
We now have `MockRateLimiter` and `MockPlatformUtils`. I didn't mock the `Dispatcher` yet (or use the existing mock for that)..